Plugins/BuildServerIntegration/AzureDevOpsIntegration/AzureDevOpsAdapter.cs

```diff
@@ -144,97 +145,132 @@ private async Task ObserveBuildsAsync(DateTime? sinceDate, bool running, IObserv
+                    foreach (Build? build in builds)
+                    {
+                        observer.OnNext(CreateBuildInfo(build));
                     }
-                    catch (UnauthorizedAccessException)
+                }
+                else
+                {
+                    // Display cached builds results
+                    if (!sinceDate.HasValue && _buildsCache.FinishedBuilds.Any())
                     {
-                        string errorMessage = _badTokenErrorMessage.Text;
-                        ManageError(errorMessage);
-
-                        if (ProjectOnErrorKey is null || ProjectOnErrorKey != CacheKey)
+                        foreach (BuildInfo? buildInfo in _buildsCache.FinishedBuilds)
                         {
-                            ProjectOnErrorKey = CacheKey;
-
-                            TaskDialogButton btnOpenSettings = new("Open settings");
-                            TaskDialogButton btnIgnore = new("Ignore");
-                            TaskDialogPage page = new()
-                            {
-                                Heading = errorMessage,
-                                Icon = TaskDialogIcon.Error,
-                                AllowCancel = true,
-                                Caption = _buildIntegrationErrorCaption.Text,
-                                Buttons = { btnOpenSettings, btnIgnore }
-                            };
-
-                            TaskDialogButton result = TaskDialog.ShowDialog(page);
-                            if (result == btnOpenSettings)
-                            {
-                                ProjectOnErrorKey = null;
-                                Validates.NotNull(_openSettings);
-                                _openSettings.Invoke();
-                            }
+                            observer.OnNext(buildInfo);
                         }
 
-                        return;
+                        // Reduce the scope to query only the builds finished after the last one in the cache
+                        sinceDate = _buildsCache.LastCall;
                     }
-                    catch (Exception ex)
-                    {
-                        string errorMessage = _genericErrorMessage.Text + ex.Message;
-                        ManageError(errorMessage);
 
-                        if (ProjectOnErrorKey is null || ProjectOnErrorKey != CacheKey)
+                    IEnumerable<Build>? builds = await _apiClient.QueryFinishedBuildsAsync(_buildDefinitions, sinceDate);
+                    IEnumerable<IGrouping<string?, Build>>? buildsByCommit = builds.GroupBy(b => b.SourceVersion);
+
+                    foreach (IGrouping<string?, Build>? buildsForACommit in buildsByCommit)
+                    {
+                        Build? buildToDisplay = buildsForACommit.OrderByDescending(b => b.FinishTime).First();
```

Comment:
I see below you're using `MinBy(...)`. Wouldn't it work here too?

Reply:
`By`s methods are equivalent of `.FirstOrDefault()` so I prefer to keep it like that.